src/librbd/operation/FlattenRequest.cc

```diff
@@ -111,70 +112,66 @@ void FlattenRequest<I>::send_op() {
+bool FlattenRequest<I>::send_detach_child() {
   I &image_ctx = this->m_image_ctx;
   assert(image_ctx.owner_lock.is_locked());
   CephContext *cct = image_ctx.cct;
 
-  ldout(cct, 5) << this << " send_update_header" << dendl;
-  m_state = STATE_UPDATE_HEADER;
-
   // should have been canceled prior to releasing lock
   assert(image_ctx.exclusive_lock == nullptr ||
          image_ctx.exclusive_lock->is_lock_owner());
 
+  // if there are no snaps, remove from the children object as well
+  // (if snapshots remain, they have their own parent info, and the child
+  // will be removed when the last snap goes away)
   {
-    RWLock::RLocker parent_locker(image_ctx.parent_lock);
-    // stop early if the parent went away - it just means
-    // another flatten finished first, so this one is useless.
-    if (!image_ctx.parent) {
-      ldout(cct, 5) << "image already flattened" << dendl;
-      return true;
+    RWLock::RLocker snap_locker(image_ctx.snap_lock);
+    if ((image_ctx.features & RBD_FEATURE_DEEP_FLATTEN) == 0 &&
+        !image_ctx.snaps.empty()) {
+      return send_update_header();
     }
-    m_parent_spec = image_ctx.parent_md.spec;
   }
-  m_ignore_enoent = true;
 
-  // remove parent from this (base) image
-  librados::ObjectWriteOperation op;
-  cls_client::remove_parent(&op);
+  ldout(cct, 2) << "detaching child" << dendl;
+  m_state = STATE_DETACH_CHILD;
 
-  librados::AioCompletion *rados_completion = this->create_callback_completion();
-  int r = image_ctx.md_ctx.aio_operate(image_ctx.header_oid,
-        				 rados_completion, &op);
-  assert(r == 0);
-  rados_completion->release();
+  auto req = image::DetachChildRequest<I>::create(
+    image_ctx, this->create_callback_context());
+  req->send();
   return false;
 }
 
 template <typename I>
-bool FlattenRequest<I>::send_update_children() {
+bool FlattenRequest<I>::send_update_header() {
   I &image_ctx = this->m_image_ctx;
   assert(image_ctx.owner_lock.is_locked());
   CephContext *cct = image_ctx.cct;
 
+  ldout(cct, 5) << this << " send_update_header" << dendl;
+  m_state = STATE_UPDATE_HEADER;
+
   // should have been canceled prior to releasing lock
   assert(image_ctx.exclusive_lock == nullptr ||
          image_ctx.exclusive_lock->is_lock_owner());
 
-  // if there are no snaps, remove from the children object as well
-  // (if snapshots remain, they have their own parent info, and the child
-  // will be removed when the last snap goes away)
-  RWLock::RLocker snap_locker(image_ctx.snap_lock);
-  if ((image_ctx.features & RBD_FEATURE_DEEP_FLATTEN) == 0 &&
-      !image_ctx.snaps.empty()) {
-    return true;
+  {
+    RWLock::RLocker parent_locker(image_ctx.parent_lock);
+    // stop early if the parent went away - it just means
+    // another flatten finished first, so this one is useless.
+    if (!image_ctx.parent) {
+      ldout(cct, 5) << "image already flattened" << dendl;
+      return true;
+    }
+    m_parent_spec = image_ctx.parent_md.spec;
```

Comment:
@dillaman `m_parent_spec` looks like unused any more and can be removed.

And just a note (not sure it needs fixing), the patch "librbd: flatten operation should detach child first" is correct only if it is applied together with "librbd: flatten request now handles clone v2", because when you swap `send_update_header` and `send_update_children`, `m_parent_spec` is initialized (in `send_update_header`) after it is used (in `send_update_children`).